buildSrc/src/main/java/org/elasticsearch/gradle/ReaperService.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.gradle;
+
+import org.gradle.api.GradleException;
+import org.gradle.api.logging.Logger;
+import org.gradle.internal.jvm.Jvm;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+public class ReaperService {
+
+    private Logger logger;
+    private Path buildDir;
+    private Path inputDir;
+    private Path logFile;
+    private volatile Process reaperProcess;
+
+    public ReaperService(Logger logger, Path buildDir, Path inputDir) {
+        this.logger = logger;
+        this.buildDir = buildDir;
+        this.inputDir = inputDir;
+        this.logFile = inputDir.resolve("reaper.log");
+    }
+
+    /**
+     * Register a pid that will be killed by the reaper.
+     */
+    public void registerPid(String serviceId, long pid) {
+        String[] killPidCommand = OS.<String[]>conditional()
+            .onWindows(() -> new String[]{"Taskill", "/F", "/PID", String.valueOf(pid)})
+            .onUnix(() -> new String[]{"kill", "-9", String.valueOf(pid)})
+            .supply();
+        registerCommand(serviceId, killPidCommand);
+    }
+
+    /**
+     * Register a system command that will be run by the reaper.
+     */
+    public void registerCommand(String serviceId, String... command) {
```

Comment:
This is public because I will need it in a future PR for registering vagrant halt commands. I only added the pid method here as a convenience because I thought it would be a fairly common thing to do. This method is the real intended api, the pid method is just sugar.

Reply:
I concur. Let's make this `private` for now, and if future processes need some other mechanism to be reaped beyond `kill` we can bake that into the plugin in a safe and robust way.

Reply:
What would making this private (and thus requiring all types of services the reaper can work on be statically known) achieve? I see this as equivalent to if we made the java's that could be used with javaexec static. This is the API we are providing for the reaper, and consumers (other plugins) use it to register system commands they would run themselves. As I mentioned above, I will be using this in a followup PR for ensuring vagrant is shut down. I don't see any reason the reaper knowing about vagrant would make this any more robust. A command would still have to be generated, somewhere, whether here, in the reaper itself, or by the plugin using the reaper. In the last case, the logic for how to deal with vagrant stays with the plugin that started vagrant, which I think is much better than having some magic reaper that happens to know how to shutdown a handful of services.

Reply:
I don't have strong feelings here. If the API is "here's a general way to run arbitrary commands" then that's fine. As alpar said, we just have to ensure that usages are good about ensuring these things are quick and relatively robust.

Reply:
I don't have strong feelings either, it just seems a bit too broad.
E.x. for vagrant having something like a public `registerLinuxCommand` that asserts that we are running on Linux might make more sense so we fail fast if accidentally registering vagrant shutdown commands   on Windows.
I'm also fine to leave it like ti is.